delivery/services/dds_service.py

```diff
@@ -0,0 +1,148 @@
+import json
+from tornado import gen
+
+from delivery.models.db_models import StagingStatus, DeliveryStatus
+from delivery.exceptions import ProjectNotFoundException, TooManyProjectsFound, InvalidStatusException
+
+log = logging.getLogger(__name__)
+
+
+class DDSService(object):
+
+    def __init__(self, external_program_service, staging_service, delivery_repo, session_factory, dds_conf):
+        self.external_program_service = external_program_service
+        self.mover_external_program_service = self.external_program_service
+        self.staging_service = staging_service
+        self.delivery_repo = delivery_repo
+        self.session_factory = session_factory
+        self.dds_conf = dds_conf
+
+    @staticmethod
+    @gen.coroutine
+    def _run_mover(delivery_order_id, delivery_order_repo, external_program_service, session_factory, dds_conf):
+        session = session_factory()
+
+        # This is a somewhat hacky work-around to the problem that objects created in one
+        # thread, and thus associated with another session cannot be accessed by another
+        # thread, therefore it is re-materialized in here...
+        delivery_order = delivery_order_repo.get_delivery_order_by_id(delivery_order_id, session)
+        try:
+            cmd = [
+                    'dds',
+                    '-tp', dds_conf["token_path"],
+                    '-l', dds_conf["log_path"],
+                    'data', 'put',
+                    '--source', delivery_order.delivery_source,
+                    '-p', delivery_order.dds_project_id,
+                    '--silent',
+                    ]
+
+            log.debug("Running dds with cmd: {}".format(" ".join(cmd)))
+
+            execution = external_program_service.run(cmd)
+            delivery_order.delivery_status = DeliveryStatus.delivery_in_progress
+            delivery_order.mover_pid = execution.pid
+            session.commit()
+
+            execution_result = yield external_program_service.wait_for_execution(execution)
+
+            if execution_result.status_code == 0:
+                delivery_order.delivery_status = DeliveryStatus.delivery_successful
+                log.info(f"Successfully delivered: {delivery_order}")
+            else:
+                delivery_order.delivery_status = DeliveryStatus.delivery_failed
+                error_msg = f"Failed to deliver: {delivery_order}. DDS returned status code: {execution_result.status_code}"
+                log.error(error_msg)
+                raise RuntimeError(error_msg)
+
+        except Exception as e:
+            delivery_order.delivery_status = DeliveryStatus.delivery_failed
+            log.error(f"Failed to deliver: {delivery_order} because this exception was logged: {e}")
+            raise e
+        finally:
+            # Always commit the state change to the database
+            session.commit()
+
+    @staticmethod
+    @gen.coroutine
+    def _get_dds_project_id(delivery_project, external_program_service):
```

Comment:
I might be missing something here, but the `delivery_project` should be the project ID already, so I don't think we need this function at all. 
The tricky thing is that mover had a `delivery_project_id` (example: delivery123456) as well as a `mover_delivery_id` (example: AB-1234-ngi2016001-123456789). The first was the delivery ID visible for the user and the latter was an internal one needed to query status from mover, but in dds we only need the delivery_project (example: snpseq00003). The `project_name` corresponds to our internal project IDs (example: AB-1234).

So If I'm not mistaken, we wouldn't even have to update the `delivery_orders` table because the `delivery_project` column is what we need.

Reply:
Aha, I thought `delivery_project` was our internal project id. I'll review my changes and update :+1: